### ISSUE TITLE
KeyboardEvent: Unexpected undefined value of key on autocomplete

### ISSUE THREAD

**Issue by NurNils** [microsoft/TypeScript#59631]
### 🔎 Search Terms

"keyboardevent", "keyboardevent key property", "keyboardevent key undefined"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about KeyboardEvent

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/PTAEBUHtQJwUwA4BsCWBjAhgFzqLALXAIwFcBzUAM0iSUgHc9CBnXZnBZgLgCgRQAjADpQAQQAm40NiwwUpHKABEGElkhpIAW2RwcSvNAwA7AJ6gAPCmMI1oYAD4qKOEimRjTNihx8wAJhEAMQ0SZi9QEgRxbDgpKxs7Rz9QAGYRAGFUNABrUA9pM1BmcjI4dhQPABppShwYegwYcXCCXBy4U3EGTzgANzhjLFAUcMoUeCl6H3xCyONxOHHjONAO0xrmazRcNvhQRvDjaDC4GBGhs4w0LErPaYII9aJIJqkfViRKHh5NY3Y1p1uvRjAAJEziJBnUAAXlAAAoAN6A8wAXy4oAA0p0Xm8AKIDIYASlhTkRPFAlNAf2YNDgQjoZHhAHJ1syausiTxUT9pgsGEIMJICYMsAAZUY4FYwFnrYHGdko+XghZQmBcnhAA

### 💻 Code

```ts
// 1. Add attribute "autocomplete" to any <input /> field on the site
// 2. Focus the updated <input />
// 3. Click on any suggestion,

const keydownHandler = ({ key }: KeyboardEvent) => {
    console.log('key', key)
}

window.addEventListener('keydown', keydownHandler)


```

### 🙁 Actual behavior

The keydown event is triggered after applying any suggestion from autocomplete with an undefined key in KeyDownEvent.

### 🙂 Expected behavior

Since the KeyboardEvent objects describe a user interaction with the keyboard, no keydown interaction should be triggered on applying an autocomplete suggestion. If the keydown event should be fired, then the key prop can be undefined and needs to be adjusted anyway.

**Comment by snarbies** [comment]
I'm not seeing this behavior in firefox. Some googling shows the behavior you describe exists in chrome.

When and how the browser fires events is beyond the purview of typescript. Typescript just type-checks your code. It has no influence on actual runtime (mis)behavior.

**Comment by MartinJohns** [comment]
TypeScript does not control when those events are triggered.

You could make the argument that the type should be `string | undefined` instead, but according to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) `undefined` should not even happen (but I can confirm that it does happen when selecting a suggestion).

Related: #31136

**Comment by RyanCavanaugh** [comment]
Chromium considers this to be working as intended, but it seems to be an obvious spec violation. We don't really have a policy for what to do if a browser vendor is intentionally violating the W3C spec that's used to generate the lib.dom.d.ts. See https://issues.chromium.org/issues/40539270

**Comment by PeytonSch** [comment]
Currently working to keep this from crashing my app as well. Related chromium issue from 2018 that is still open: https://issues.chromium.org/issues/41425904 

If you want to replicate easily:

```html
<!DOCTYPE html>
<html>

<body>
    <div>
        <h2>No form</h2>
        <input id="no-form" type="text" autocomplete="off" />
    </div>
    <div>
        <h2>Form</h2>
        <form>
            <input id="form" type="text" autocomplete="on" />
            <input type="submit" value="Send Request" />
        </form>
    </div>
    <script>
        document.body.addEventListener('keydown', (e) => {
            console.log(e);
            console.log(e instanceof KeyboardEvent);
            console.log(e.keyCode);
            console.log(e.key.toLowerCase());
        });
    </script>
</body>

</html>
```

If you want to crash your react app:

```ts
const KeyDownTest = () => {

    useEffect(() => {
        const handleKeyDown = (e: KeyboardEvent) => {
            console.log(e instanceof KeyboardEvent);
            console.log(e);
            console.log(e.key.toLowerCase());
        };

        window.addEventListener('keydown', handleKeyDown);

        return () => {
            window.removeEventListener('keydown', handleKeyDown);

        };
    }, []); 

    return (
        <>
            <div>
                <h2>No form</h2>
                <input id="no-form" type="text" autoComplete="off" />
            </div>
            <div>
                <h2>Form</h2>
                <form>
                    <input id="form" type="text" autoComplete="on" />
                    <input type="submit" value="Send Request" />
                </form>
            </div>
        </>
    )
}

export default KeyDownTest
```

I am trying to hack my way around what MartinJohns pointed out with having the type for the key be `string | undefined`.

For example:
```ts
            const x: string | undefined = undefined;
            console.log(x.toLowerCase());    // x possibly undefined
```
This fails to compile with the error that x is possibly undefined, catching this potential issue early and preventing runtime crashes.

**Comment by RyanCavanaugh** [comment]
Thanks for the link @PeytonSch . Closing as external
